### ISSUE TITLE
What are the pairs for the JSON string to generate a module?

### ISSUE THREAD

**Issue by AmirSimantov** [drush-ops/drush#4590]
The help for the generate command states:

**Options:
 --answers=ANSWERS   A JSON string containing pairs of question and answers.**

However, I cannot figure out **what are the keys** for this string and could not find any documentation for it.

I would like to create a module with no interactions, passing all values as arguments.

Any idea? An example would also be appreciated. Thanks!

**Comment by Sweetchuck** [comment]
This is a problem for all the "generate" command, not just for the create new module command.
Only with CLI it is impossible to find out the correct structure of the JSON for the --answers option. There is no such a "help" or "description".

However for custom generator command one can use the "::$description" or "::getDescription()" to give such a hint for the user.

But there are some cases where a static description is not enough, for example when a question depends on the previous answer.

---

**How many modules should be generated?**
>2

**Module 1 machine-name?**
> my_module_01


**Module 2 machine-name?**
> my_module_02

---

I just playing with generators so far, nothing serious yet, but this is why I don't use the `::collectVars()`, and the question string contains the key of the answer. Like this
```php
<?php
$question = new Question(
  'my_key_01 - Answer to the Ultimate Question of Life, the Universe, and Everything?',
  '42'
);
$this->vars['my_key_01'] = $this->ask($input, $output, $question)
```

Then I add an extra confirmation question:
```php
<?php
$json_encode_flags = \JSON_UNESCAPED_UNICODE | \JSON_UNESCAPED_SLASHES | \JSON_PRETTY_PRINT;
$question = new ConfirmationQuestion(
  'Do you want to continue with the following values?' . PHP_EOL . json_encode($this->vars, $json_encode_flags),
  TRUE,
);

$this->vars['go'] = $this->ask($input, $output, $question);
```
I know that this is far from the optimal, but at least the user can reproduce the same JSON string for the --answer option.

**Comment by MurzNN** [comment]
You can dump all question keys (and separately optional questions list) via little hack of generator module code. For example, to collect list of keys for `drush generate module` - open file `vendor/chi-teck/drupal-code-generator/src/Command/Drupal_8/Module/Standard.php` and in function `interact()` add the dump string:
```php
print_r(array_keys($questions)); die;
```
Full example:
```php
  protected function interact(InputInterface $input, OutputInterface $output) {
    $questions = Utils::defaultQuestions();
    $questions['description'] = new Question('Module description', 'The description.');
    $questions['package'] = new Question('Package', 'Custom');
    $questions['dependencies'] = new Question('Dependencies (comma separated)');
    print_r(array_keys($questions)); die;
```
and output:
```
Array
(
    [0] => name
    [1] => machine_name
    [2] => description
    [3] => package
    [4] => dependencies
)
```
Same for optional questions:
```php
    print_r(array_keys($option_questions)); die;
```

**Comment by MurzNN** [comment]
And even easier and more universal way - is to dump those values in function `collectVars()` of file `vendor/chi-teck/drupal-code-generator/src/Helper/InputHandler.php`:
```
  public function collectVars(InputInterface $input, OutputInterface $output, array $questions, array $vars = []) {
    $output->writeln('Question group keys: ' . implode(', ', array_keys($questions)));
```

**Comment by weitzman** [comment]
This is really a feature request for DCG. Cant be solved in Drush. /cc @Chi-teck

**Comment by Chi-teck** [comment]
That's really hard to guess those keys without looking into DCG source code. I encourage you to use Drush 11 which ships with DCG 2. In this case the anwers option is no long accepts JSON but a sequence of answers. Like follows.
```
$ drush generate module --answer=Example --answer=example

 Welcome to module generator!
––––––––––––––––––––––––––––––

 Module name [Docroot]:
 ➤ Example

 Module machine name [example]:
 ➤ example

 Module description [Provides additional functionality for the site.]:
 ➤ 

```
So keys are not needed anymore. The provided anwsers are printed to STDOUT so it's quite easy to figure out what answers are expected.  Also it supports `--dry-run` option and provides some debug information in `-vv` mode.
Note that DCG 1 will reach EOL same day as Drush 10.

**Comment by Chi-teck** [comment]
There is one more option that works for all Drush versions, actually for any CLI application built with Symfony Console. You can feed interctive commands using STDIN stream like this.
`echo -e 'Example\nexample\nThe module description\n' | drush generate module`

**Comment by weitzman** [comment]
What is the extra info with -vv? If it is the stuff below, I only see that with -vvv. Note that I fixed #4373 yesterday.

```
[debug] Collected variables:
(
    [machine_name] => sut
    [name] => Sut
    [class] => SutController
    [services] => Array
        (
        )

    [route_name] => sut.example
    [route_path] => /sut/example
    [route_title] => Example
    [route_permission] => access content
)

**Comment by Chi-teck** [comment]
Yes, that's what I meant. It's displayed with `-vv` in 11.0.0-RC3.
